@here/harp-mapview/lib/MapView.ts

```diff
@@ -2518,6 +2556,94 @@ export class MapView extends THREE.EventDispatcher {
         return this.m_elevationProvider;
     }
 
+    private extractAttitude() {
+        const camera = this.m_camera;
+        const projection = this.projection;
+
+        const cameraPos = cache.vector3[1];
+        const transform = cache.transform[0];
+        const tangentSpaceMatrix = cache.matrix4[1];
+        // 1. Build the matrix of the tangent space of the camera.
+        cameraPos.setFromMatrixPosition(camera.matrixWorld); // Ensure using world position.
+        projection.localTangentSpace(projection.unprojectPoint(cameraPos), transform);
+        tangentSpaceMatrix.makeBasis(transform.xAxis, transform.yAxis, transform.zAxis);
+
+        // 2. Change the basis of matrixWorld to the tangent space to get the new base axes.
+        cache.matrix4[0].getInverse(tangentSpaceMatrix).multiply(camera.matrixWorld);
+        transform.xAxis.setFromMatrixColumn(cache.matrix4[0], 0);
+        transform.yAxis.setFromMatrixColumn(cache.matrix4[0], 1);
+        transform.zAxis.setFromMatrixColumn(cache.matrix4[0], 2);
+
+        // 3. Deduce orientation from the base axes.
+        let yaw = 0;
+        let pitch = 0;
+        let roll = 0;
+
+        // Decompose rotation matrix into Z0 X Z1 Euler angles.
+        const epsilon = 1e-10;
+        const d = transform.zAxis.dot(cameraPos.set(0, 0, 1));
+        if (d < 1.0 - epsilon) {
+            if (d > -1.0 + epsilon) {
+                yaw = Math.atan2(transform.zAxis.x, -transform.zAxis.y);
+                pitch = Math.acos(transform.zAxis.z);
+                roll = Math.atan2(transform.xAxis.x, transform.yAxis.z);
+            } else {
+                // Looking bottom-up with space.z.z == -1.0
+                yaw = -Math.atan2(-transform.yAxis.x, transform.xAxis.x);
+                pitch = 180;
+                roll = 0;
+            }
+        } else {
+            // Looking top-down with space.z.z == 1.0
+            yaw = Math.atan2(-transform.yAxis.x, transform.xAxis.x);
+            pitch = 0.0;
+            roll = 0.0;
+        }
+
+        return {
+            yaw,
+            pitch,
+            roll
+        };
+    }
+
+    private lookAtImpl(params: LookAtParams): void {
+        const zoomLevel = THREE.MathUtils.clamp(
+            getOptionValue(params.zoomLevel, this.zoomLevel),
+            this.m_minZoomLevel,
+            this.m_maxZoomLevel
+        );
+        const tilt = Math.min(getOptionValue(params.tilt, this.tilt), PITCH_LIMIT_DEG);
+        const target = GeoCoordinates.fromObject(getOptionValue(params.target, this.target));
```

Comment:
Applies to `tilt` also.

We're calling `extractAttidute` 3 times for each call to `lookAt` (by always reading `this.tilt`, `this.heading` and `updateLookAtSettings`.

Can we
 1) cache `extractAttitude` output ?
 2) at least call read current values only if they are really needed: s/getOptionValue/good old ?: operator/

Reply:
As mentioned above: In a follow up change I will cache titl and heading (or yaw, pitch , roll) in the `ViewState`